src/scenarios/scenario_perception_pipeline.cpp

```diff
@@ -0,0 +1,156 @@
+  {
+    bool is_successful = sendTargetPose(pose_stamped.pose);
+    if (is_successful)
+      success_number++;
+    else
+      failure_number++;
+  }
+
+  return { success_number, failure_number };
+}
+
+bool ScenarioPerceptionPipeline::sendTargetPose(const geometry_msgs::msg::Pose& target_pose)
+{
+  move_group_interface_ptr_->setPoseTarget(target_pose);
+
+  moveit::planning_interface::MoveGroupInterface::Plan plan_msg;
+  if (move_group_interface_ptr_->plan(plan_msg) == moveit::core::MoveItErrorCode::SUCCESS)
+  {
+    move_group_interface_ptr_->execute(plan_msg);
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+ScenarioPerceptionPipelineFixture::ScenarioPerceptionPipelineFixture()
+{
+  createTestCases();
+}
+
+void ScenarioPerceptionPipelineFixture::SetUp(::benchmark::State& /*state*/)
+{
+  if (node_.use_count() == 0)
+  {
+    node_ = std::make_shared<rclcpp::Node>("test_scenario_perception_pipeline_node",
+                                           rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true));
+
+    node_->get_parameter("selected_test_case_index", selected_test_case_index_);
+  }
+
+  if (move_group_interface_ptr_.use_count() == 0)
+  {
+    move_group_interface_ptr_ = std::make_shared<MoveGroupInterface>(node_, PLANNING_GROUP);
+  }
+}
+
+void ScenarioPerceptionPipelineFixture::TearDown(::benchmark::State& /*state*/)
+{
+}
+
+void ScenarioPerceptionPipelineFixture::createTestCases()
+{
+  readTestCasesFromFile(TEST_CASES_YAML_FILE);
+}
+
+nav_msgs::msg::Path ScenarioPerceptionPipelineFixture::selectTestCase(size_t test_case_index)
+{
+  return test_cases_.at(test_case_index);
+}
+
+void ScenarioPerceptionPipelineFixture::readTestCasesFromFile(const std::string& yaml_file_name)
+{
+  YAML::Node config = YAML::LoadFile(yaml_file_name.c_str());
+  for (YAML::const_iterator it = config["test_cases"].begin(); it != config["test_cases"].end(); ++it)
+  {
+    const std::string yaml_string = dynmsg::yaml_to_string(*it);
+    nav_msgs::msg::Path test_case = getTestCaseFromYamlString(yaml_string);
+
+    test_cases_.push_back(test_case);
+  }
+}
+
+nav_msgs::msg::Path ScenarioPerceptionPipelineFixture::getTestCaseFromYamlString(const std::string& yaml_string)
+{
+  nav_msgs::msg::Path path_msg;
+  void* ros_message = reinterpret_cast<void*>(&path_msg);
+
+  dynmsg::cpp::yaml_and_typeinfo_to_rosmsg(dynmsg::cpp::get_type_info({ "nav_msgs", "Path" }), yaml_string, ros_message);
+
+  return path_msg;
+}
+
+BENCHMARK_DEFINE_F(ScenarioPerceptionPipelineFixture, test_scenario_perception_pipeline)(benchmark::State& st)
+{
+  auto selected_test_case = selectTestCase(selected_test_case_index_);
+  for (auto _ : st)
+  {
+    auto sc = ScenarioPerceptionPipeline(move_group_interface_ptr_);
+    auto [success_number, failure_number] = sc.runTestCase(selected_test_case);
+    st.counters["success_number"] = success_number;
+    st.counters["failure_number"] = failure_number;
```

Comment:
I am confused, where is the elapsed time measured?

Reply:
Google benchmark library handles this stuff automatically. `st.counters` is used optionally. If you are still confused, ping me again.

Reply:
Thanks!